### PR TITLE
server: Return nullopt when process_context is nullptr

### PR DESCRIPTION
**Pull request by justin-mp**
The InstanceImpl::processContext() function returns a ProcessContextOptRef
a.k.a. absl::optional<std::reference_wrapper<ProcessContext>>.  When
InstanceImpl::process_context_ is nullptr, the returned value an
absl::optional with a value of std::reference_wrapper<>(nullptr),
better known as an illegal reference to nullptr.  While everything is
fine in the InstanceImpl, anyone who tries to use the returned
reference will be greeted with a crash.

This commit follows the lead of the initializer of InstanceImpl::api_
and returns absl::nullopt when InstanceImpl::process_context_ is
nullptr.

Risk: Low
Testing: bazel test //test/server:server_test
Documentation: N/A
Release Notes: N/A

Signed-off-by: Justin Mazzola Paluska <justinmp@google.com>

### CODE REVIEW

File: test/server/server_test.cc [hunk 1]
```diff
@@ -1411,6 +1411,28 @@ TEST_P(ServerInstanceImplTest, DisabledExtension) {
...
+  initialize("test/server/test_data/server/empty_bootstrap.yaml");
+  ProcessContextOptRef context = server_->processContext();
+  EXPECT_FALSE(context.has_value());
+
+  // Prior to the commit when this test was added, the code would return a
+  // context where has_value() was true, producing an opt ref that has a value
+  // which is a reference pointing to null. Doing anything on that reference
+  // would cause a crash. The rest of this test is ensuring that case doesn't
+  // occur again.
+  if (context.has_value()) {
```
Comment (zuercher) [inline test/server/server_test.cc:1428]: I think it's fine to remove everything after the `EXPECT_FALSE`. That will cause the test to fail if this code incorrectly returns a non-empty option.

File: source/server/server.cc [hunk 2]
```diff
@@ -246,6 +246,14 @@ void InstanceImpl::flushStatsInternal() {
 
 bool InstanceImpl::healthCheckFailed() { return !live_.load(); }
 
+ProcessContextOptRef InstanceImpl::processContext() {
+  if (process_context_ == nullptr) {
+    return absl::nullopt;
+  } else {
```
Comment (zuercher) [inline source/server/server.cc:252]: nit: No need for the else since the if block returns.

File: test/server/server_test.cc (hunk 1 above)
Comment (justin-mp) [reply test/server/server_test.cc]: Done.

File: source/server/server.cc (hunk 2 above)
Comment (justin-mp) [reply source/server/server.cc]: Done.

**Review by zuercher — APPROVED** [review]

**Pull request merged**